ENIGMAsystem/SHELL/Graphics_Systems/OpenGL3/GL3vertex.cpp

```diff
@@ -15,39 +17,191 @@
+    vertexBuffer->dirty = false;
+  } else {
+    bind_array_buffer(enigma::vertexBufferPeers[buffer]);
   }
 
-  bool vertex_format_exists(int id){
-    return (id >= 0 && (unsigned)id < enigma::vertexFormats.size() &&  enigma::vertexFormats[id] != nullptr);
+  //(enigma::vertexBufferPeers[buffer]);
+
+  enigma::glsl_attribute_enable_all_internal(false); //Disable all attributes
+
+  bool useTextCoords = false, useColors = false;
+  size_t offset = 0;
+  map<int,int> useCount;
+  const size_t stride = vertexFormat->stride * sizeof(float);
+  for (size_t i = 0; i < vertexFormat->flags.size(); ++i) {
+    const pair<int, int> flag = vertexFormat->flags[i];
+
+    size_t elements = 0, size = 0;
+    GLenum type = GL_FLOAT;
+    switch (flag.first) {
+      case vertex_type_float1: elements = 1; size = 1; break;
+      case vertex_type_float2: elements = 2; size = 2; break;
+      case vertex_type_float3: elements = 3; size = 3; break;
+      case vertex_type_float4: elements = 4; size = 4; break;
+      case vertex_type_color: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
+      case vertex_type_ubyte4: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
+    }
+
+    if (flag.second == vertex_usage_color) useColors = true;
+    if (flag.second == vertex_usage_textcoord) useTextCoords = true;
+
+    // NOTE: This is not what GMS1.4 does, it uses glBindAttribLocation
+    // so that in_Color0 is an alias of in_Color which glBindAttribLocation allows
+    // GMS1.4 also apparently does this binding before the shaders are linked
+    // according to a GMS tech blog:
+    // https://www.yoyogames.com/blog/16/shaders-overview-part-2
+    // https://www.yoyogames.com/blog/18/shaders-overview-part-4
+    std::string name = "";
+    int usageIndex = useCount[flag.second]++;
+    switch (flag.second) {
+      case vertex_usage_position: name = "in_Position"; break;
+      case vertex_usage_color: name = "in_Color"; break;
+      case vertex_usage_normal: name = "in_Normal"; break;
+      case vertex_usage_textcoord: name = "in_TextureCoord"; break;
+    }
+    if (!name.empty()) {
+      int location = enigma::graphics_find_attribute_location(name, usageIndex);
+      if (location == -1 && flag.second == vertex_usage_color) {
+        location = enigma::graphics_find_attribute_location("in_Colour", usageIndex);
+      }
+
+      if (location != -1) {
+        enigma::glsl_attribute_enable_internal(location, true);
+        enigma::glsl_attribute_set_internal(location, elements, type, (type == GL_UNSIGNED_BYTE), stride, offset);
+      }
+    }
+
+    offset += size;
   }
 
-  void vertex_format_add(int id, int type, int attribute) {
-    enigma::vertexFormats[id]->AddAttribute(type, attribute);
+  enigma::glsl_uniformf_internal(enigma::shaderprograms[enigma::bound_shader]->uni_color, (float)enigma::currentcolor[0]/255.0f, (float)enigma::currentcolor[1]/255.0f, (float)enigma::currentcolor[2]/255.0f, (float)enigma::currentcolor[3]/255.0f);
```

Comment:
static_cast

Reply:
That does nothing but add verbosity.

Reply:
I copied and pasted Harri's code there from the model class.